storage refactor to address issues in #65

ikreymer opened the pull request:
- use Symbols instead of direct properties to avoid custom proxied objects showing up in getOwnPropertyNames(), (though still show up in getOwnPropertySymbols())
- proxy override fixes: add ownKeys() override to fix Object.keys() only listing objects in storage
- ensure internal properties are not accessible/visible directly -- only visible through getOwnPropertySymbols() and getOwnPropertyDescriptors() which are less likely to be used
- add toString() override
- tests: add override storage tests with proxy, incorporating most of tests from #66 but with proxy object applied

ikreymer commented:
@ibnesayeed I think this addresses most of the issues you've found!

I've incorporated the tests from #66 into a separate `overrides-storage` tests which test `localStorage` and `sessionStorage` objects with the proxy applied, rather than with creating a `new Storage()`, so this tests the proxy object is working as expected.

The custom properties are unfortunately still accessible via `Object.getOwnPropertySymbols(localStorage)` or `Object.getOwnPropertyDescriptors(localStorage)` however these are less likely to be used in normal course of using local/sessionStorage than 
`Object.getOwnPropertyNames()` or `Object.keys()`, which should list just the contents of the storage as expected.

ibnesayeed commented:
This looks good to me. Using Symbols was something I was thinking to suggest, but I did not see it being used anywhere else Wombat (other than overrides), so I thought you might have a reason to avoid that, but I am glad that we are leveraging that to mask the visibility.

I am wondering, do we really need to keep the data inside a nested property? What benefits does it bring? While other attributes that we are trying to hide can continue toe be put under symbol-properties, if we surface the user added data properties directly, we might be able to avoid a bunch or accessor overrides.

> I've incorporated the tests from #66 into a separate `overrides-storage` tests which test `localStorage` and `sessionStorage` objects with the proxy applied, rather than with creating a `new Storage()`, so this tests the proxy object is working as expected.

I was thinking, should the conditional proxy be applied directly in the main storage implementation, because I am not sure where would be useful to have the storage object without accessor-based getter/setter. Though, i will leave that up to you as you have better understanding of the whole system while I investigated only specific parts of the code. Also, please feel free to amend #66 (or let me know if you want me to make any changes there) to remove any test cases that you have already incorporated elsewhere before merging the remaining (if necessary).

ikreymer commented:
> I am wondering, do we really need to keep the data inside a nested property? What benefits does it bring? While other attributes that we are trying to hide can continue toe be put under symbol-properties, if we surface the user added data properties directly, we might be able to avoid a bunch or accessor overrides.

Yes, you're right, I've now simplified it to remove the nested DATA symbol. That eliminated some of the accessor overrides.
However, the Proxy object is still needed to intercept setters and getters in order to fire events when a property is changed via accessors or via setItem.

> > I've incorporated the tests from #66 into a separate `overrides-storage` tests which test `localStorage` and `sessionStorage` objects with the proxy applied, rather than with creating a `new Storage()`, so this tests the proxy object is working as expected.
> 
> I was thinking, should the conditional proxy be applied directly in the main storage implementation, because I am not sure where would be useful to have the storage object without accessor-based getter/setter. Though, i will leave that up to you as you have better understanding of the whole system while I investigated only specific parts of the code. Also, please feel free to amend #66 (or let me know if you want me to make any changes there) to remove any test cases that you have already incorporated elsewhere before merging the remaining (if necessary).

Right, yeah, that's why I split off these tests into their own file, working with the existing storage. The direct tests can probably be removed altogether, if the event fire testing is also included in the override tests.
The direct tests don't really represent actual usage since storage is never created with `new Storage()` for example.
The actual wombat just calls `createStorage()` which returns an object wrapped in a Proxy always (unless proxies aren't available).

I think this PR supersedes #66 since I migrated most of the tests, can probably close that one. I suppose those tests could be refactored to use `createStorage()` but at that point, might as well use the full override imo.
The only thing that might make sense to add is event firing tests to the overrides, I didn't have a chance to do that.

ibnesayeed commented:
I really liked these new changes including removal of data nesting and relocation of the storage creation wrapper in the storage implementation file.

> Right, yeah, that's why I split off these tests into their own file, working with the existing storage. The direct tests can probably be removed altogether, if the event fire testing is also included in the override tests.

I was thinking along the lines of consolidating all the storage tests in the original storage test file itself. Because, now we have two storage related test files, which will eventually get subsumed into the new one, essentially renaming the old file with the new one with more tests in it. Not that it matters! :laughing:
